### ISSUE TITLE
False positive for TS 2367 when value is changed inside a function

### ISSUE THREAD

**Issue by ezsh** [microsoft/TypeScript#39893]
**TypeScript Version:**  Nightly

**Search Terms:**

**Expected behavior:** No errors reported

**Actual behavior:** TS reports the error

**Related Issues:**

**Code**
```ts
type T = "foo" | "bar";

class A {
    value: T;
    constructor(v: T) {
        this.value = v;
    }
}

let a = new A("foo");

function f(a: A) {
    a.value = "bar";
}

if (a.value === "foo") {
    f(a);
    if (a.value === "bar") { // no errors are expected here
        console.log("bar!");
    }
}
```

<details><summary><b>Output</b></summary>

```ts
"use strict";
class A {
    constructor(v) {
        this.value = v;
    }
}
let a = new A("foo");
function f(a) {
    a.value = "bar";
}
if (a.value === "foo") {
    f(a);
    if (a.value === "bar") {
        console.log("bar!");
    }
}

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "useDefineForClassFields": false,
    "alwaysStrict": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "downlevelIteration": false,
    "noEmitHelpers": false,
    "noLib": false,
    "noStrictGenericChecks": false,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "esModuleInterop": true,
    "preserveConstEnums": false,
    "removeComments": false,
    "skipLibCheck": false,
    "checkJs": false,
    "allowJs": false,
    "declaration": true,
    "experimentalDecorators": false,
    "emitDecoratorMetadata": false,
    "target": "ES2017",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play/index.html?ts=Nightly#code/C4TwDgpgBAKlC8UBEAzA9mpUA+yBGAhgE5IDcAsAFBUDGANgQM6NQCCUA3lVD1AG4E6AVwgAuWBUq8oNNADtGwIkJrA0RABR9xMAJSdu03sAAWAS0YA6AcOiI+k6QF8qL6pToRgUAgihyIAHc2DVQMJF1JKhQhOVUzeSgUDQJxVn0uKV4Ca0ERPyRCEkk3KjMUKBTc2wR4RDDMDMMeZIJI5qhyypybfLr6ooiDLKMZeUY0T0s6NABzUKKAQgjHXjcnIA)

**Comment by ezsh** [comment]
The following gives the same error:
```ts
type T = "foo" | "bar";
let v: T = "foo";

function g(): void {
    v = "bar";
}

if (v === "foo") {
    g();
    if (v === "bar") { // false positive 2367 here, console prints out "bar!"
        console.log("bar!");
    }
}
```

**Comment by IllusionMH** [comment]
See #9998

**Comment by ezsh** [comment]
Thanks!
